File: Sum_root_to_leaf_numbers_recursive.java
// Time Complexity : O(N)
// Space Complexity : O(H), where H is height of tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class TreeNode1 {
	int val;
	TreeNode1 left;
	TreeNode1 right;

	TreeNode1() {
	}

	TreeNode1(int val) {
		this.val = val;
	}

	TreeNode1(int val, TreeNode1 left, TreeNode1 right) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}

public class Sum_root_to_leaf_numbers_recursive {
	public int sumNumbers(TreeNode1 root) {
		return helper(root, 0);
	}

	public int helper(TreeNode1 root, int curSum) {
		// base case
		if (root == null) {
			return 0;
		}
		// logic
		if (root.left == null && root.right == null) {
			return curSum * 10 + root.val;
		}
		int left = helper(root.left, curSum * 10 + root.val);
		int right = helper(root.right, curSum * 10 + root.val);
		return left + right;
	}
}

File: sum_root_to_leaf_numbers_recursive.py
"""
Problem: 129. Sum Root to Leaf Numbers
Leetcode: https://leetcode.com/problems/sum-root-to-leaf-numbers/
Solution: Recursive DFS preorder traversal
Time Complexity: O(n), we have to traverse all numbers
Space Complexity: O(h), h is the tree height, recursive approach uses inbuilt stack to store all numbers
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    total_sum = 0

    def sumNumbers(self, root: TreeNode) -> int:
        self.helper(root, self.total_sum)
        return self.total_sum

    def helper(self, node, curr_sum):
        # Base case
        if not node:
            return

        # Leaf node
        if not node.left and not node.right:
            # calculate the total sum
            self.total_sum = self.total_sum + curr_sum * 10 + node.val
            return

        curr_sum = curr_sum * 10 + node.val
        # calculate the left subtree
        self.helper(node.left, curr_sum)
        # calculate the right subtree
        self.helper(node.right, curr_sum)